README.md
# Sven Internal
Sven Internal is a C++ plugin for [SvenMod](https://github.com/sw1ft747/SvenMod) that provides to you various cheats and gameplay enhances

Currently supported version of the game: 5.25

Imagine cheating in a co-op game =)

# How to install
First, if you don't have installed SvenMod then download it and install (see [readme](https://github.com/sw1ft747/svenmod)). Download the archive `svenint.rar` from ([releases](https://github.com/sw1ft747/sven_internal/releases)) and place all files from the archive in the root folder of the game. Next, add the plugin (`sven_internal.dll`) to the file `plugins.txt` (see the header `Adding plugins` in SvenMod's [readme](https://github.com/sw1ft747/svenmod)). 

# Features
- Menu (key **INSERT** as default)
- Customizable Configs
- ESP
- Aimbot
- No Recoil
- Wallhack
- Glow & Chams
- Vectorial Strafer
- Auto Bunnyhop
- Auto Jumpbug
- Fast Run
- Fake Lag
- Color Pulsator
- Speed Hack
- Anti-AFK
- Cam Hack
- Key Spammer
- First-Person Roaming
- Various Visual Hacks (Velometer, Crosshair, etc.)
- Message Spammer
- Models Manager
- Skybox Replacement
- Custom Vote Popup
- Custom Chat Colors

# Files of the plugin
The plugin uses subfolder `sven_internal` in the root directory of the game.

How it looks: `../Sven Co-op/sven_internal/`.

This folder is used to save the config, load list of players (their Steam64 ID) for **Chat Colors** and load spam tasks for **Message Spammer**.

Folder `config` contains all config files. You can save your config via menu or console command `sc_save_config`, also you can load a config via concommand `sc_load_config <optional: filename>`. Automatically the plugin load the config named `default.ini`.

File `chat_colors_players.txt` allows to change the color of nickname for a specific player, will be automatically loaded at plugin load.

Folder `message_spammer` is used by **Message Spammer** to load spam tasks.

Also, when cheat loaded it will execute `sven_internal.cfg` file from folder `../Sven Co-op/svencoop/`.

# Console Variables/Commands
Type in the console the following command: `sm printcvars all ? sc_`.

The command above will print information about each CVar/ConCommand that belongs to the plugin.

# Chat Colors
That feature lets you change the color of nickname of players when they write something in the chat.

File `chat_colors_players.txt` used for adding the players in such format: `STEAM64ID : COLOR_NUMBER`.

To convert traditional SteamID to Steam64 ID faster, you can use the following console command: `sc_steamid_to_steam64id [SteamID]`.

There're currently 5 color numbers (slots) plus rainbow color, thus you can use 6 unique and customizable colors (can be changed in Menu).

Rainbow color has slot 0, others custom colors have slotos from 1 to 5.

Example for the file:
```
76561198819023292 : 0 # it's a comment
76561197962091295 : 5 ; it's a comment too!
```

File `chat_colors_players.txt` automatically loads when cheat loaded. Also, you can use the console command `sc_chat_colors_load_players` to reload the list of players.

# Models Manager
Lets you change the models of players the working directory is `models_manager`

File `random_models.txt` contains the list of models that will be randomly used

File `target_players.txt` contains the list of pairs of type `STEAM64ID = MODELNAME` to replace models for specified players

File `ignored_players.txt` contains the list of Steam64 ID's that will be ignored when you replace models for specified players

The files above will be automatically loaded when the plugin is loaded

For more information, check the files above, they contain comments

# Message Spammer
Roughly, it's some kind of AHK.

It uses `*.txt` files from folder `message_spammer` to run spam tasks. Use the following console command to run a spam task: `sc_ms_add <filename>`.

It supports 3 keywords: `loop`, `send` and `sleep`.

Important: Message Spammer reads the `*.txt` files sequentially.

For example:
```
sleep 1.5
send test string
```
It will be executed like: wait 1.5 seconds, then send to game chat `test string`

### loop
Must be placed at the beginning of the file.

Loops the spam task, otherwise if you don't place that keyword in your .txt file, then the spam task will be executed once.

### send [message]
Sends a message to game chat.

One argument: [message].

### sleep [delay]
Sleeps a spam task.

One argument: [delay].

**Example:**
```
loop
send stuck
sleep 0.35
send play
sleep 0.35
send diff
sleep 0.35
send alone
sleep 0.35
send light
sleep 0.35
send my ears
sleep 0.35
send vote
sleep 0.35
send buy
sleep 0.35
send thief
sleep 125.0
```


changelog.txt
Changelog for Sven Internal

24/07/2022 - Version 2.0.13:
- Added entity list
- Added snap lines in ESP
- Added aimbot, silent aimbot and ragebot [for a while it doesn't target hitboxes] (experimental)
- Added console commands to toggle Chams & Glow (sc_chams, sc_glow)
- Added support of wrench when use Fast Crowbar
- Added several models in class table
- Enhanced render of the box in ESP (you can disable it in the menu for better performance) [Menu > Visual > ESP > Optimize]
- Fixed size of the buffer in the module for drawing

20/07/2022 - Version 2.0.12:
- Added point strafing (sc_strafe_dir 4, sc_strafe_point_x <coordinate>, sc_strafe_point_y <coordinate>)
- Added strafer for Stick feature (sc_stick, sc_stick_strafe)
- Added Models Manager to change models for players (experimental)
- Added Human Grunt Medic & Torch in the class table
- Added styles for ESP, changed the font of ESP, added feature to disable ESP through walls (for players)

18/07/2022 - Version 2.0.11:
- Some fixes in the code
- Added feature No Recoil (experimental)
- Added missing CVars: 'setang' and 'getang'

15/07/2022 - Version 2.0.10:
- Added auto edgejump
- Added some missing commands (sc_strafe_ignore_ground, sc_spam_hold_mode, sc_rotate_dead_body, sc_tertiary_attack_glitch, sc_quake_guns, sc_first_person_roaming)
- Fixed Chams & Glow when enabled first-person roaming
- Moved all menu's stuff to draw it to the class CMenuModule

11/07/2022 - Version 2.0.9:
- Minor fixes
- Updated for newer version of SvenMod
- Added feature to ignore different versions of the same maps [Menu > Utility > Misc.] (won't work on completely different maps)
- Used class 'CMessageBuffer' instead of poor namespace 'UserMessages'
- Returned check of slopes for jumpbug
- Console command 'sc_prints_steamids' won't print empty Steam64 IDs anymore

08/07/2022 - Version 2.0.8:
- Added one tick exploit (sc_one_tick_exploit)
- Added some models in class table
- Added prediction of hotizontal position in auto jumpbug
- Now auto jumpbug takes in account the server's gravity
- Fast Crowbar, Fast Medkit features moved to the menu: Player > Misc.
- Added the second method of Fast Crowbar that freezes the player (sc_fastcrowbar2)

07/07/2022 - Version 2.0.7:
- Minor fixes

06/07/2022 - Version 2.0.6:
- Added the missing parameter "AutoCeilClipping" in the config
- Added rapid fire on crowbar and medkit (sc_fastcrowbar, sc_fastmedkit) [Thanks to guwi]
- Added feature to register commands that will be executed each game tick (sc_register_on_tick_command, sc_remove_on_tick_command, sc_print_on_tick_commands)

06/06/2022 - Version 2.0.5:
- Fixed Anti-AFK when player is loading on the map with enabled option to stay within radius
- Fixed duplicate call of function 'WindowStyle' in Menu module
- Added 'sc_stick <player_index>' command to stick to a specified player
- Added some They Hunger models in class table
- Adjusted fall velocity when jumpbug is used

09/04/2022 - Version 2.0.4:
- IMPROTANT: place your old config "../sven_internal/sven_internal.ini" in this folder (create if needed) "../sven_internal/config/sven_internal.ini" and rename it to "default.ini"
- Folders used by Sven Internal will be created automatically
- Now ESP shows players from other teams as enemies
- Enhacned Anti-AFK mode "Spam Gibme": to don't lose weapons, before suicide, player will select a weapon from 1st slot in the inventory
- Added Anti-AFK mode "Spam Kill" same as "Spam Gibme"
- Added simple manager of configs that lets you to save/load/remove/create any count of configs in the Menu
- To the console command "sc_load_config" can be passed optional argument to load a specific config (for example, sc_load_config my_config)
- Fixed gauss' tertiary attack glitch when use spinner
- Fixed issues with CamHack when enabled spinner
- Fixed inability to use longjump with spinner

27/03/2022 - Version 2.0.3:
- Added BunnymodXT-like speedometer
- Added feature to automatically hide HUD when use CamHack
- Added feature to change the colors of HUD (experimental)

26/03/2022 - Version 2.0.2:
- Added new mode to draw entities: draw players hitboxes only
- Updated CamHack, now it doesn't depend on user's command inputs
- Updated method to disable mouse inputs when the menu is opened
- Used PM's member "maxspeed" instead of "clientmaxspeed"
- Removed pointless conversion from radians to degrees when processing jumpbug

21/03/2022 - Version 2.0.1:
- Added missing console command "sc_test"
- Added check in Spinner if player wants to fire the weapon
- Added Assert macro in visual.h > ProcessBones

20/03/2022 - Version 2.0.0:
- Reworked as a plugin for SvenMod

patterns.cpp
#include "patterns.h"

namespace Patterns
{
	namespace Hardware
	{
		DEFINE_PATTERN(flNextCmdTime, "D9 1D ? ? ? ? 75 0A A1");

		DEFINE_PATTERN(Netchan_CanPacket, "D9 05 ? ? ? ? D9 EE DA E9 DF E0 F6 C4 44 8B 44 24 04");
		DEFINE_PATTERN(Netchan_Transmit, "B8 ? ? ? ? E8 ? ? ? ? A1 ? ? ? ? 33 C4 89 84 24 ? ? ? ? 8B 84 24 ? ? ? ? 53 55");

		DEFINE_PATTERN(Sys_InitializeGameDLL, "E8 ? ? ? ? 33 C0 83 3D ? ? ? ? ? 0F 9F C0 50 E8 ? ? ? ? 83 C4");

		DEFINE_PATTERN(SCR_UpdateScreen, "83 EC 44 A1 ? ? ? ? 33 C4 89 44 24 40 83 3D ? ? ? ? ? 0F 85 ? ? ? ? C7");
		DEFINE_PATTERN(SCR_BeginLoadingPlaque, "6A ? E8 ? ? ? ? A1 ? ? ? ? 83 C4 ? 83 F8");
		DEFINE_PATTERN(SCR_EndLoadingPlaque, "C7 05 ? ? ? ? ? ? ? ? E8 ? ? ? ? 68 ? ? ? ? 68 ? ? ? ? E8 ? ? ? ? 83 C4 ? C7 05");

		DEFINE_PATTERN(V_RenderView, "81 EC ? ? 00 00 A1 ? ? ? ? 33 C4 89 84 24 ? ? 00 00 D9 EE D9 15");
		DEFINE_PATTERN(R_RenderScene, "83 EC ? E8 ? ? ? ? 85 C0 74");
		DEFINE_PATTERN(R_RenderScene_HOOKED, "E9 ? ? ? ? ? ? ? 85 C0 74");
		DEFINE_PATTERN(R_SetupFrame, "83 EC 24 A1 ? ? ? ? 33 C4 89 44 24 20 33 C0 83 3D ? ? ? ? 01 0F 9F C0 50 E8 ? ? ? ? E8");
		DEFINE_PATTERN(R_LoadSkyboxInt, "81 EC ? ? 00 00 A1 ? ? ? ? 33 C4 89 84 24 ? ? 00 00 53 8B 9C 24 ? ? 00 00 55 56 57 89 5C 24 24");

		DEFINE_PATTERN(CL_ClearState, "E8 ? ? ? ? 85 C0 75 ? FF 74 24 04");
		DEFINE_PATTERN(CL_ParseConsistencyInfo, "83 EC ? A1 ? ? ? ? 33 C4 89 44 24 24 D9 ? ? ? ? ? D9 ? DA");
		DEFINE_PATTERN(CL_BatchResourceRequest, "55 8B EC 83 E4 ? B8 ? ? ? ? E8 ? ? ? ? A1 ? ? ? ? 33 C4 89 84 24 1C 01 04 00");
		DEFINE_PATTERN(CL_PlayerFlashlight, "81 EC ? ? ? ? A1 ? ? ? ? 33 C4 89 84 24 BC 00 00 00 A1");

		DEFINE_PATTERN(IsSafeFileToDownload, "81 EC ? ? ? ? A1 ? ? ? ? 33 C4 89 84 24 04 01 00 00 8B 84 24 0C 01 00 00 56 85 C0");
		
		DEFINE_PATTERN(CRC_MapFile, "81 EC ? ? ? ? A1 ? ? ? ? 33 C4 89 84 24 84 04 00 00");
		DEFINE_PATTERN(ResourceList_Offset, "A1 ? ? ? ? 85 C0 74 ? 83 3D ? ? ? ? ? 74 ? 89 86 84 00 00 00");
		DEFINE_PATTERN(UserInfo_Offset, "8D B8 ? ? ? ? 75 20 8B 0D");
		DEFINE_PATTERN(clc_buffer, "68 ? ? ? ? 85 F6 74 ? 0F AE E8");
	}

	namespace Client
	{
		DEFINE_PATTERN(IN_Move, "83 3D ? ? ? ? 00 75 21 83 3D ? ? ? ? 00 74 18 FF 74 24 08 F3 0F 10 44 24 08 51 F3 0F 11 04 24");

		DEFINE_PATTERN(CVotePopup__MsgFunc_VoteMenu, "56 57 FF 74 24 10 8B F1 FF 74 24 18 E8 ? ? ? ? E8 ? ? ? ? 89 86 ? ? 00 00 E8");
		DEFINE_PATTERN(READ_BYTE, "8B 0D ? ? ? ? 8D 51 01 3B 15 ? ? ? ? 7E 0E C7 05 ? ? ? ? 01 00 00 00 83 C8 FF C3 A1 ? ? ? ? 89 15 ? ? ? ? 0F B6 04 08 C3");
		DEFINE_PATTERN(READ_STRING, "8B 15 ? ? ? ? 33 C0 53 55 8B 2D ? ? ? ? 56 57 8B 3D ? ? ? ? C6 05 ? ? ? ? 00 90");
		
		DEFINE_PATTERN(CHudBaseTextBlock__Print, "55 8B EC 6A FF 68 ? ? ? ? 64 A1 00 00 00 00 50 53 56 57 A1 ? ? ? ? 33 C5 50 8D 45 F4 64 A3 00 00 00 00 8B D9 8B 0D");

		DEFINE_PATTERN(CVoiceBanMgr__SaveState, "81 EC ? ? 00 00 A1 ? ? ? ? 33 C4 89 84 24 ? ? 00 00 8B 84 24 ? ? 00 00 53 57 FF 35 ? ? ? ? 8B F9 50 8D 44 24 14");
		DEFINE_PATTERN(CVoiceBanMgr__SetPlayerBan, "56 FF 74 24 08 8B F1 E8 ? ? ? ? 80 7C 24 0C 00 74 13 85 C0 75 32 FF 74 24 08 8B CE E8");
		DEFINE_PATTERN(CVoiceBanMgr__InternalFindPlayerSquelch, "53 55 8B 6C 24 0C 56 57 0F 10 4D 00 0F 28 C1 66 0F 73 D8 08 66 0F FC C8 0F 10 C1 66 0F 73 D8 04");

		DEFINE_PATTERN(CVoiceStatus__IsPlayerBlocked, "83 EC 14 A1 ? ? ? ? 33 C4 89 44 24 10 56 8D 44 24 04 8B F1 50 FF 74 24 20 FF 15");
		DEFINE_PATTERN(CVoiceStatus__SetPlayerBlockedState, "81 EC ? ? 00 00 A1 ? ? ? ? 33 C4 89 84 24 ? ? 00 00 53 68 ? ? ? ? 8B D9 FF 15 ? ? ? ? D9 5C 24 08");
		DEFINE_PATTERN(CVoiceStatus__UpdateServerState, "81 EC ? ? 00 00 A1 ? ? ? ? 33 C4 89 84 24 ? ? 00 00 53 8B D9 89 5C 24 08");

		DEFINE_PATTERN(CClient_SoundEngine__Play2DSound, "83 EC ? F3 0F 10 44 24 14");
		DEFINE_PATTERN(CClient_SoundEngine__LoadSoundList, "81 EC ? ? ? ? A1 ? ? ? ? 33 C4 89 84 24 A8 08 00 00");

		DEFINE_PATTERN(HACK_GetPlayerUniqueID, "FF 74 24 08 FF 74 24 08 FF 15 ? ? ? ? 83 C4 08 85 C0 0F 95 C0 C3");

		DEFINE_PATTERN(GetClientColor, "8B 4C 24 04 85 C9 7E 35 6B C1 ? 0F BF 80 ? ? ? ? 48 83 F8 03");

		DEFINE_PATTERN(ScaleColors, "66 0F 6E 4C 24 10 8B 4C 24 04 0F 5B C9");
		DEFINE_PATTERN(ScaleColors_RGBA, "8B 4C 24 04 0F B6 41 03 66 0F 6E C8 0F B6 01 0F 5B C9");

		DEFINE_PATTERN(CHud__Think, "83 EC 08 53 56 57 8B F9 E8 ? ? ? ? 8B 37 D9 5C 24");

		DEFINE_PATTERN(V_PunchAxis, "8B 44 24 04 F3 0F 10 44 24 08 F3 0F 11 04 85");

		DEFINE_PATTERN(WeaponsResource__GetFirstPos, "6B 54 24 04 68 56 57 33 F6 8B F9 81 C2 ? ? ? ? 8B 02 85 C0 74");
	}

	namespace Server
	{
		DEFINE_PATTERN(gpGlobals, "A1 ? ? ? ? 8B 80 98 00 00 00 03 86 80 00 00 00 50 68 ? ? ? ? 6A ? FF 15 ? ? ? ? 83 C4");
	}

	namespace GameOverlay
	{
		DEFINE_PATTERN(SetCursorPos_Hook, "55 8B EC 80 3D ? ? ? ? 00 74 19 8B 45 08 A3 ? ? ? ? 8B 45 0C A3 ? ? ? ? B8 01");
		DEFINE_PATTERN(ValveUnhookFunc, "55 8B EC C7");
		//DEFINE_PATTERN(ValveUnhookFunc, "55 8B ? 64 ? ? ? ? ? 6A ? 68 ? ? ? ? 50 64 ? ? ? ? ? ? 81 ? ? ? ? ? 56 8B ? ? 85");
	}
}